pkg/controller/certificaterequest/cloudflare.go

```diff
@@ -49,59 +49,84 @@ type CloudflareResponse struct {
+			// If the SOA TTL exceeds this, give up immediately.
+			if sleepDuration > maxNegativeCacheTTL {
+				reqLogger.Info("negative cache TTL is too long; giving up")
+				return false
+			}
+			// If the TTL is shorter than the default wait time, disregard.
+			if ttl > sleepDuration {
+				sleepDuration = ttl
+			}
 		}
 
+		reqLogger.Info(fmt.Sprintf("attempt %v to verify resource record %v has been updated with value %v", attempt, fqdn, txtValue))
 		attempt++
-	}
 
-	errMsg := fmt.Sprintf("unable to verify that resource record %v has been updated to value %v after %v attempts.", fqdn, txtValue, maxAttemptsForDnsPropogationCheck)
-	reqLogger.Error(errors.New(errMsg), errMsg)
-	return false
-}
+		reqLogger.Info(fmt.Sprintf("will query DNS in %v seconds", sleepDuration))
+		time.Sleep(time.Duration(sleepDuration) * time.Second)
 
-// verifyDnsResourceRecordUpdate will wait the waitTimePeriodDnsPropogationCheck
-// and then check if the DNS changes have propagated
-func verifyDnsResourceRecordUpdate(reqLogger logr.Logger, fqdn string, txtValue string) bool {
-	reqLogger.Info(fmt.Sprintf("will query DNS in %v seconds", waitTimePeriodDnsPropogationCheck))
+		response = new(CloudflareResponse)
 
-	time.Sleep(time.Duration(waitTimePeriodDnsPropogationCheck) * time.Second)
+		err = FetchResourceRecordUsingCloudflareDNS(reqLogger, fqdn, response)
+		if err != nil {
+			reqLogger.Error(err, "failed to fetch DNS records")
+			continue
+		}
 
-	dnsChangesPropogated, err := ValidateResourceRecordUpdatesUsingCloudflareDNS(reqLogger, fqdn, txtValue)
-	if err != nil {
-		reqLogger.Error(err, "could not validate DNS propagation.")
-		return false
+		// If there is no answer field, this is likely an expected NXDOMAIN response.
+		if len(response.Answers) == 0 {
+			reqLogger.Info("no answers received from cloudflare; likely not propagated")
+			continue
+		}
+
+		// Trim any trailing dot from the answer name and quotes from the data.
+		cfName := strings.TrimSuffix(response.Answers[0].Name, ".")
+		cfData := strings.Trim(response.Answers[0].Data, "\"")
+
+		if strings.EqualFold(cfName, fqdn) && cfData == txtValue {
+			return true
+		}
+
+		reqLogger.Info("could not validate DNS propagation for " + fqdn)
 	}
 
-	return dnsChangesPropogated
+	errMsg := fmt.Sprintf("unable to verify that resource record %v has been updated to value %v after %v attempts.", fqdn, txtValue, maxAttemptsForDnsPropagationCheck)
+	reqLogger.Error(errors.New(errMsg), errMsg)
+	return false
 }
 
-// ValidateResourceRecordUpdatesUsingCloudflareDNS contacts cloudflareDnsOverHttpsEndpoint and validates the json response.
-func ValidateResourceRecordUpdatesUsingCloudflareDNS(reqLogger logr.Logger, name string, value string) (bool, error) {
+// FetchResourceRecordUsingCloudflareDNS contacts cloudflareDnsOverHttpsEndpoint and returns the json response.
+func FetchResourceRecordUsingCloudflareDNS(reqLogger logr.Logger, name string, outResponse *CloudflareResponse) error {
```

Comment:
Instead of mutating the input response inside each time, should we just return `(*CloudflareResponse, error) from this?

Reply:
Since `VerifyDnsResourceRecordUpdate` carries the response struct into its next loop iteration to check for a negative cache result, I thought it would be _slightly_ clearer for `VerifyDnsResourceRecordUpdate` to own the struct.  (Maybe not relevant; still coming at it from a C mindset.)